torch/csrc/autograd/python_variable.cpp

```diff
@@ -252,16 +252,7 @@ int THPVariable_set_grad(THPVariable *self, PyObject *py_grad)
       "can't assign Variable as its own grad");
 
   auto& grad = ((THPVariable*)py_grad)->cdata;
-  bool gradIsSparse = false;
-  auto backend = var.is_cuda() ? Backend::SparseCUDA : Backend::SparseCPU;
-  auto typeOpt = at::globalContext().getNonVariableTypeOpt(backend, var.scalar_type());
-  if (typeOpt) {
-       auto& sparseType = at::globalContext().getNonVariableType(backend, var.scalar_type());
-       auto& gradType = at::globalContext().getNonVariableType(grad.type().backend(), grad.scalar_type());
-       gradIsSparse = gradType == sparseType;
-  }
-
-  THPUtils_assertRet(-1, grad.type() == var.type() || gradIsSparse,
+  THPUtils_assertRet(-1, grad.type() == var.type() || grad.is_sparse(),
```

Comment:
unless I'm missing something, this doesn't look equivalent?  The new code allows you to assign a sparse gradient, even if it's the wrong dtype.